### PR TITLE
KAA-531: Collect information about amount of active endpoint sessions and system load average for LB function

### PR DESCRIPTION
**Pull request by ashvayka**

### CODE REVIEW

File: server/control/src/main/java/org/kaaproject/kaa/server/control/service/loadmgmt/dynamicmgmt/EndpointCountRebalancer.java [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        int minLoadedOpsServer = getMinLoadedOpsServer(opsServerLoadHistory);
+        int minEndpointCount = getLastEndpointCount(opsServerLoadHistory.get(minLoadedOpsServer));
+        int maxLoadedOpsServer = getMaxLoadedOpsServer(opsServerLoadHistory);
+        int maxEndpointCount = getLastEndpointCount(opsServerLoadHistory.get(maxLoadedOpsServer));
+        LOG.info("Min loaded ops server is {} with {} endpoints", minLoadedOpsServer, minEndpointCount);
+        LOG.info("Max loaded ops server is {} with {} endpoints", maxLoadedOpsServer, maxEndpointCount);
+        int maxDiff = maxEndpointCount - minEndpointCount;
+        LOG.info("Max difference between endpoint counts is {}", maxDiff);
+        if (maxDiff < minDiff) {
+            LOG.debug("Max endpoint count difference is to small to trigger recalculation. Min required diff is {}", minDiff);
```
Comment (batytskyy) [inline server/control/src/main/java/org/kaaproject/kaa/server/control/service/loadmgmt/dynamicmgmt/EndpointCountRebalancer.java]: Please correct to small -> too small

File: server/control/src/main/java/org/kaaproject/kaa/server/control/service/loadmgmt/dynamicmgmt/EndpointCountRebalancer.java [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+            if (curWeight >= 0) {
+                LOG.debug("No redirection rules for {}", targetLoad);
+                continue;
+            }
+            List<RedirectionRule> redirectionRules = new ArrayList<RedirectionRule>();
+            for (Entry<Integer, Double> targetWeight : weights.entrySet()) {
+                if (targetWeight.getValue() < 0) {
+                    continue;
+                }
+                double initRedirectProbability = Math.abs(curWeight) * targetWeight.getValue() * maxInitRedirectProbability;
```
Comment (batytskyy) [inline server/control/src/main/java/org/kaaproject/kaa/server/control/service/loadmgmt/dynamicmgmt/EndpointCountRebalancer.java:114]: Where does this formula come from?

File: server/common/zk/src/main/avro/zk.avsc [hunk 3]
```diff
@@ -14,7 +14,8 @@
  "type": "record",
  "name": "LoadInfo",
  "fields": [
-     {"name": "loadIndex",  "type": "int"}
+     {"name": "endpointCount",  "type": "int"},
+     {"name": "loadAvarage",  "type": "double"}
```
Comment (batytskyy) [inline server/common/zk/src/main/avro/zk.avsc]: Please correct spelling to loadAverage

File: server/operations/src/main/java/org/kaaproject/kaa/server/operations/service/akka/actors/core/OperationsServerActor.java [hunk 4]
```diff
@@ -49,6 +55,8 @@
     /** The tenants id-actor map. */
     private final Map<String, ActorRef> tenants;
 
+    private Map<UUID, StatusRequestState> statusRequestStatesMap = new HashMap<UUID, StatusRequestState>();
```
Comment (batytskyy) [inline server/operations/src/main/java/org/kaaproject/kaa/server/operations/service/akka/actors/core/OperationsServerActor.java]: Please make this variable final and initialize it in the constructor.

File: server/operations/src/main/java/org/kaaproject/kaa/server/operations/service/akka/actors/core/TenantActor.java [hunk 5]
```diff
@@ -67,6 +74,8 @@
 
     private final String tenantId;
 
+    private Map<UUID, StatusRequestState> statusRequestStatesMap = new HashMap<UUID, StatusRequestState>();
```
Comment (batytskyy) [inline server/operations/src/main/java/org/kaaproject/kaa/server/operations/service/akka/actors/core/TenantActor.java]: Please make this variable final and initialize it in the constructor.

File: server/operations/src/main/java/org/kaaproject/kaa/server/operations/service/loadbalance/DefaultLoadBalancingService.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2014 CyberVision, Inc.
```
Comment (batytskyy) [inline server/operations/src/main/java/org/kaaproject/kaa/server/operations/service/loadbalance/DefaultLoadBalancingService.java]: Please correct license header: 2014  -> 2014-2015

File: server/operations/src/main/java/org/kaaproject/kaa/server/operations/service/loadbalance/LoadBalancingService.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright 2014 CyberVision, Inc.
```
Comment (batytskyy) [inline server/operations/src/main/java/org/kaaproject/kaa/server/operations/service/loadbalance/LoadBalancingService.java]: Please correct license header: 2014  -> 2014-2015

**Comment by batytskyy** [comment]
Minor comments, otherwise LGTM.

File: server/control/src/main/java/org/kaaproject/kaa/server/control/service/loadmgmt/dynamicmgmt/EndpointCountRebalancer.java (hunk 2 above)
Comment (ashvayka) [inline server/control/src/main/java/org/kaaproject/kaa/server/control/service/loadmgmt/dynamicmgmt/EndpointCountRebalancer.java:114]: Redirect probability is based on three values:
1. Initial configuration parameter.
2. Weight of the endpoint that is overloaded
3. Weight of the endpoint that is not enough loaded.

**Pull request merged**